### PR TITLE
[SYCL][ESIMD] Implement unified memory API - atomic_update(acc,...) zero operands

### PR DESCRIPTION
**Pull request by sarnex**
This implements the compile-time properties version of atomic_update with accessors and zero operands.

I moved the zero operand block to be before the one operand block, so hopefully it's easier to review. I didn't port the local accessor ones yet, I'll do that as part of the SLM change.

I also moved the one and two operand intrinsics to supported even though they aren't used by and supported code yet, it's just preparing for the next change.

I also fixed some typos in the usm version doc.

### CODE REVIEW

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp [hunk 1]
```diff
@@ -518,74 +633,104 @@ bool test_fp_types(queue q, const Config &cfg) {
...
-  passed &= test<float, N, Op, UseMask, UsePVCFeatures>(q, cfg);
+  passed &= run_test<UseAcc, float, N, Op, UseMask, UsePVCFeatures>(q, cfg);
 #ifndef CMPXCHG_TEST
-  if (q.get_device().has(sycl::aspect::atomic64) &&
-      q.get_device().has(sycl::aspect::fp64)) {
-    passed &= test<double, N, Op, UseMask, UsePVCFeatures>(q, cfg);
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+  constexpr bool test_double = true;
+#else
+  constexpr bool test_double = !UseAcc;
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp:646]: Why double is not tested in acc+stateful mode?

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp [hunk 2]
```diff
@@ -475,40 +561,69 @@ struct ImplFcmpwr : ImplCmpxchgBase<T, N, atomic_op, atomic_op::fcmpxchg> {};
...
-          bool UsePVCFeatures, int SignMask = (Signed | Unsigned)>
+          bool UsePVCFeatures, bool UseAcc, int SignMask = (Signed | Unsigned)>
 bool test_int_types(queue q, const Config &cfg) {
   bool passed = true;
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+  constexpr bool test_long = true;
+  constexpr bool test_short = true;
+#else
+  constexpr bool test_long = !UseAcc;
+  constexpr bool test_short = !UseAcc;
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp:582]: Are short an long supported only in usm atomic update?

**Review by v-klochkov — COMMENTED** [review]
I finished the 1st code-review pass.
2 major questions: 
* why short/long/double are not tested in acc+stateful mode?
* 'predec' may need either deprecation/removal or additional updates in atomic_update comments and implementation.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -4518,6 +4570,319 @@ atomic_update(Tx *p, Toffset byte_offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+/// atomic_update(AccessorT acc, simd<Toffset, N> byte_offset,
+///               simd_mask<N> mask, props = {});               /// (acc-au0-1)
+///
+/// Atomically updates \c N memory locations represented by an accessor and
+/// a vector of offsets, and returns a vector of old values found at the
+/// memory locations before update. The update operation has no arguments
+/// in addition to the value at the memory location.
+///
+/// @tparam Op The atomic operation - can be \c atomic_op::inc or
+/// \c atomic_op::dec, \c atomic_op::load.
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:4601]: Looks like predec has one argument, see [here](https://github.com/intel/llvm/blob/156b730a6886b1394b7b91e2455d6e6b42ac6ec1/sycl/include/sycl/ext/intel/esimd/common.hpp#L317)

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -4518,6 +4570,319 @@ atomic_update(Tx *p, Toffset byte_offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+/// @param acc The SYCL accessor.
+/// @param byte_offset The vector of 32-bit or 64-bit offsets in bytes. 64-bit
+/// offsets are supported only when stateless memory accesses are enforced, i.e.
+/// accessor based accesses are automatically converted to stateless accesses.
+/// @param mask Operation mask, only locations with non-zero in the
+///   corresponding mask element are updated.
+/// @param props The parameter 'props' specifies the optional compile-time
+///   properties list. Only L1/L2 properties are used.
+//    Other properties are ignored.
+/// @return A vector of the old values at the memory locations before the
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:4614]: doesnt look like we support predec here, see previous comment

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp (hunk 2 above)
Comment (sarnex) [reply sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp:582]: yeah same as other comment about double, only 4 byte data is supported

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp (hunk 1 above)
Comment (sarnex) [reply sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp:646]: for that intrinsic, we only support 4 byte data, and assert that. see [here](https://github.com/intel/llvm/blob/156b730a6886b1394b7b91e2455d6e6b42ac6ec1/sycl/include/sycl/ext/intel/esimd/memory.hpp#L5131)

**Review by turinevgeny — APPROVED** [review]

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp (hunk 2 above)
Comment (sarnex) [reply sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp]: i just pushed a better fix, thanks

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp [hunk 5]
```diff
@@ -302,12 +304,12 @@ bool test_acc(queue q, const Config &cfg) {
         for (int cnt = 0; cnt < cfg.repeat; ++cnt) {
           if constexpr (n_args == 0) {
             if constexpr (UseMask) {
-              if constexpr (UseProperties)
+              if constexpr (UseProperties || DataSizeRequiresLSC)
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp:307]: this never gets called if !UseProperties && DataSizeRequiresLSC, we do a constexpr return in this case earlier but the compiler still tries to instantiate these for some reason, so just make it happy

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp]: double should be supported in LSC mode. Also double as FP type is supported only in LSC mode and in both(acc and usm) variants, right?

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: i made an internal tracker to figure out whats going on with predec

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**